src/StreamingServer.php

```diff
@@ -229,22 +230,43 @@ public function handleRequest(ConnectionInterface $conn, ServerRequestInterface
             $conn->write("HTTP/1.1 100 Continue\r\n\r\n");
         }
 
+        // execute request handler callback
         $callback = $this->callback;
-        $cancel = null;
-        $promise = new Promise(function ($resolve, $reject) use ($callback, $request, &$cancel) {
-            $cancel = $callback($request);
-            $resolve($cancel);
-        });
+        try {
+            $response = $callback($request);
+        } catch (\Exception $error) {
+            // request handler callback throws an Exception
+            $response = Promise\reject($error);
+        } catch (\Throwable $error) { // @codeCoverageIgnoreStart
+            // request handler callback throws a PHP7+ Error
+            $response = Promise\reject($error); // @codeCoverageIgnoreEnd
+        }
 
         // cancel pending promise once connection closes
-        if ($cancel instanceof CancellablePromiseInterface) {
-            $conn->on('close', function () use ($cancel) {
-                $cancel->cancel();
+        if ($response instanceof CancellablePromiseInterface) {
+            $conn->on('close', function () use ($response) {
+                $response->cancel();
             });
         }
 
+        // happy path: request body is known to be empty => immediately end stream
+        if ($contentLength === 0) {
+            $stream->emit('end');
+            $stream->close();
+        }
+
+        // happy path: response returned, handle and return immediately
+        if ($response instanceof ResponseInterface) {
+            return $this->handleResponse($conn, $request, $response);
+        }
+
+        // did not return a promise? this is an error, convert into one for rejection below.
+        if (!$response instanceof PromiseInterface) {
```

Comment:
This `if` should probably omitted and promise conversion should be completely delegated to `resolve()`.

Reply:
I'm not opposed to your suggestion, but after trying this out locally, the performance improvement is much less significant. Do you feel this is more of a cosmetic issue or would this block these suggested changes?